Build mob recipes

Aliagor opened the pull request:
Added recipes for buildMob mechanic for all mobs that had craftable morbs using rustic alchemy. Note that Stymphalian Bird spawning is a bit wonky, wrote a comment in recipes.zs about it.

Krutoy242 commented:
Sounds really interesting! I will take a look.

Krutoy242 commented:
Please write what tests was done, so I don't need to test it myself.

Aliagor commented:
I tested all new buildMob recipes in both overworld and skyblock world. All works fine except for Stymphalian Bird but it is broken with morbs and spawn eggs as well. For some reason after spawning it just teleports few block below. I tried making the spawn location higher but it changes nothing, bird still goes underground or below platform in skyblock.

Krutoy242 commented:
I have a few requests for recipes.

1. Use more common ingredients.
  > In 4 years of working on E2EE I realized that nobody likes to craft something once per modpack. Ingredients like chiseled blocks or decorative blocks are a bad idea. Players often refer to such ingredients as “annoying”.
  > Here are some examples that you need to replace:  
  > ![image](https://github.com/user-attachments/assets/5f0d1bc5-4eeb-4621-a238-29ec37e7c69d)![image](https://github.com/user-attachments/assets/4622fb3d-c28f-4428-86ef-6e7aaeb72e0c)![image](https://github.com/user-attachments/assets/a6ef8468-b249-4c19-ba8c-f0fda92a0f79)


2. Fake Iron should always be used as an alternative to regular iron.
  > This is a design decision so that there is no confusion in the player's choice of block. Use Oredict here, which contains both blocks at the same time.
  > 
  > ![image](https://github.com/user-attachments/assets/2b8d9b1d-67f3-4018-b016-79420cccae62)

3. I really like the pirate design, but it's too expensive. Make it much cheaper. The most expensive block should be the Life Conglomerate.
  > The previous recipe was also very cheap. Definitely need to replace the dragon wall and head. The same applies to Dread Lich.
  > 
  > ![image](https://github.com/user-attachments/assets/e7b079d5-53b4-4a68-ab50-c506731e9676)

4. Four different ingredients out of 4 is a bad idea. Keep only 2 or even 1 block besides the conglomerate in Fallen Hero.
  > Take vanilla golem as an example - only 2 types of blocks. This also apply to Sea Serpent recipe - keep only 2-3 blocks, maybe only dragon scales (also made them oredicted to match any color)
  > 
  > ![image](https://github.com/user-attachments/assets/175a21ba-d652-4c25-bac0-8f873a4368a6)

5. Remove recipes from Rustic's Alchemy that have been replaced by construction.

Also, this isn't a directive, but I want to add a feature like this:
If the world is Peaceful, instead of spawning a mob, its items should drop when it dies. I don't yet know how to get a list of items that drop from a mob when it dies with `ZenScrip`. The easiest thing would be to refactor the `scripts\jei\entity_drop.zs` file so that it would be possible to get this data dynamically, when creating a mob in peaceful mode.
You can volunteer to do this, but if you don't want to, message me and I'll do it myself.

Aliagor commented:
Ready for review. Implemented the peaceful way for buildMob, currently if world is in peaceful it will assume that mob is hostile and drop it's loot from entity_drop.zs instead of spawning it.

Krutoy242 commented:
Thank you! Review will be delayed since im hospitalized with pneumonia for at least a week.

Krutoy242 commented on scripts/do/build/entity.zs:
```diff
@@ -139,7 +142,20 @@ zenClass MobBuild {
       val r = rotate(face, offset.x, offset.z);
       val truePos = Position3f.create(r[0] + pos.x + shiftX, offset.y + pos.y + shiftY, r[1] + pos.z + shiftZ);
-      utils.spawnGenericCreature(world, entity.id, truePos.x, truePos.y, truePos.z - 0.1, face);
+      /** TODO: Find a way to check if mob can spawn in peaceful to spawn mob instead of items */
+      if (world.getWorldInfo().difficulty == "PEACEFUL") {
```
To determine if entity is hostile and not spawn in Peaceful world, use something like `if (entity instanceof IEntityMob)` https://docs.blamejared.com/1.12/en/Vanilla/Entities/IEntityMob/

This may not work straight, since `IEntityMob` is inworld entity, after it being created.
In this case, probably `spawnGenericCreature` need to be refactored to return inworld entity.

Anyway, im suggesting using DeepSeek / ChatGPT for this - it was very handy for me. Dont forget to determine that you using `1.12.2 Forge` environment.

<img src="https://i.imgur.com/QhfFN3s.png" width=50%>

Aliagor replied:
Thanks, this helped me add detection for it.

Krutoy242 commented:
> Implemented the peaceful way for buildMob

Also, please, after notable changes or implementing features (like this peaceful mechanic) please report what tests being made, since if no tests done I will need to do that myself.

Aliagor commented on scripts/do/build/entity.zs:
```diff
@@ -19,8 +19,11 @@ import crafttweaker.util.Position3f;
 import crafttweaker.block.IBlock;
 import crafttweaker.world.IBlockPos;
 import crafttweaker.world.IWorld;
+import crafttweaker.world.IWorldInfo;
```
Removed

Aliagor commented on scripts/jei/entity_drop.zs:
```diff
@@ -32,9 +37,14 @@ function add(entity as IEntityDefinition, dropList as IItemStack[], percent as b
   } else {
     fixedList = dropList;
   }
+  entityDropTableMap[entity] = dropList as IItemStack[];
```
Removed

Aliagor commented on scripts/do/build/recipes.zs:
```diff
@@ -110,3 +112,221 @@ scripts.do.build.entity.add(<entity:thaumcraft:cultistportalgreater>, [
+  [
+    '  a  '
+  ],
+  [
```
Fixed

Aliagor commented:
I added the detection for the "hostility" of the mob. Works for all mobs except sea serpent so I added it in as exception. All different mobs that are buildable work.

Tested all available buildMob recipes in peaceful and they are working. Only exception from this is stymphalian bird because it does not have the loot defined in entity_drop.zs and is hostile. I know what items drop but I'm not sure on numbers that should go into the entity_drop.zs list of mobs.

Krutoy242 commented:
I will closely review changes later and merge.

Krutoy242 commented on scripts/do/build/recipes.zs:
```diff
@@ -110,3 +112,201 @@ scripts.do.build.entity.add(<entity:thaumcraft:cultistportalgreater>, [
+    ' b b '
+  ]
+], {
+  a: <iceandfire:dread_stone_face>,
+  b: <iceandfire:dragon_bone_wall>,
+  c: <iceandfire:dread_stone>,
+  x: <contenttweaker:conglomerate_of_life>
```
I was mentioned before - please make Dread Lich ingredients cheaper. `Dragon Bone Wall` is too complicated blocks.

Aliagor replied:
You earlier mentioned only the pirate not Dread Lich so I assumed that boss mob was fine with a bit harder blocks. Changed it to be way simpler and a bit harder than pirate.

![obraz](https://github.com/user-attachments/assets/b2ec35c7-c9eb-422c-b943-f647951ff0c3)

Krutoy242 replied:
Hehe, yes i did 
![image](https://github.com/user-attachments/assets/c55acccf-6ee6-4509-a2b6-7d837a426392)

Aliagor replied:
Ahh, nvm. I'm just blind then. :D

Krutoy242 commented on scripts/do/build/recipes.zs:
```diff
@@ -110,3 +112,201 @@ scripts.do.build.entity.add(<entity:thaumcraft:cultistportalgreater>, [
+    ' d d '
+  ]
+], {
+  a: <minecraft:skull>,
```
![image](https://github.com/user-attachments/assets/4c2e6d8c-ee0f-43ac-abba-9df0586ec297)
Skull doesn't have preview. Consider to replace it with other previewable block or find a way how to fix preview in `Compact Machines` configuration (it allow to change preview items).

Aliagor replied:
Changed skull to bone block that can render properly.

![obraz](https://github.com/user-attachments/assets/75552653-0a4b-4dbe-b6bd-a5e0b377cc24)
